File: app/src/main/java/com/hlibrary/base/handler/CommonHandler.java
package com.hlibrary.base.handler;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.IntRange;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v4.util.Pools;

import com.hlibrary.application.AppContext;
import com.hlibrary.entity.MessageEvent;
import com.hlibrary.util.Logger;
import com.hlibrary.util.ToastUtil;

import java.io.Serializable;
import java.util.concurrent.Future;


/**
 * @author linwenhui
 * @date 2015/9/15.
 */
public class CommonHandler implements HandlerInterface<CommonHandler> {

    protected static final String TAG = "CommonHandler";

    private Context context;
    protected LocalBroadcastManager localBroadcastManager;
    private String ExtraTag;
    private ProgressDialog progressDialog;
    private static Pools.SimplePool<CommonHandler> handlePool = new Pools.SimplePool<>(20);

    protected CommonHandler(Context context) {
        init(context);
    }

    protected final void init(Context context) {
        this.context = context;
        localBroadcastManager = LocalBroadcastManager.getInstance(context.getApplicationContext());
    }

    public String getExtraTag() {
        return ExtraTag;
    }

    public void setExtraTag(@NonNull String extraTag) {
        ExtraTag = extraTag;
    }

    public String getTag() {
        Logger.i(TAG, " " + TAG + " " + ExtraTag);
        if (!"".equals(ExtraTag)) {
            return ExtraTag;
        }
        return TAG;
    }

    @Override
    public void showToast(@NonNull String text) {
        ToastUtil.showLongTime(context, text);
    }

    @Override
    public void showToast(@NonNull String text, @IntRange(from = 1) int duration) {
        ToastUtil.showCustomTime(context, text, duration);
    }

    @Override
    public void showToast(@StringRes int resId) {
        ToastUtil.showLongTime(context, resId);
    }

    @Override
    public void showToast(@StringRes int resId, @IntRange(from = 1) int duration) {
        ToastUtil.showCustomTime(context, resId, duration);
    }

    @Override
    public void showProgess(@NonNull String msg) {
        if (null == progressDialog) {
            progressDialog = new ProgressDialog(context);
            progressDialog.setMessage(msg);
            progressDialog.setCancelable(false);
            progressDialog.show();
        } else {
            progressDialog.setMessage(msg);
            progressDialog.show();
        }
    }

    @Override
    public void dismissProgress() {
        if (progressDialog != null && progressDialog.isShowing())
            progressDialog.dismiss();
    }

    @Override
    public Future runNewThread(@NonNull Runnable r) {
        return AppContext.threadPool.submit(r);
    }

    @Override
    public void runOnUiThread(@NonNull Runnable r) {
        AppContext.mainHandler.post(r);
    }

    @Override
    public void runOnUiThreadForDelay(@NonNull Runnable r, @IntRange(from = 1) long delayMillis) {
        AppContext.mainHandler.postDelayed(r, delayMillis);
    }

    @Override
    public boolean sendBroadcast(@NonNull Intent intent) {
        return localBroadcastManager.sendBroadcast(intent);
    }

    @Override
    public void startActivityForCls(@NonNull Class<? extends Activity> cls) {
        context.startActivity(new Intent(context, cls));
    }

    @Override
    public void startActivityForCls(@NonNull Class<? extends Activity> cls, @NonNull Bundle options) {
        Intent intent = new Intent(context, cls);
        intent.putExtras(options);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK);
        context.startActivity(intent);
    }

    @Override
    public void startActivityForCls(@NonNull Class<? extends Activity> cls, @NonNull String value) {
        Bundle bundle = new Bundle();
        bundle.putString(getTag(), value);
        startActivityForCls(cls, bundle);
    }

    @Override
    public void startActivityForCls(@NonNull Class<? extends Activity> cls, @NonNull String[] values) {
        Bundle bundle = new Bundle();
        bundle.putStringArray(getTag(), values);
        startActivityForCls(cls, bundle);
    }

    @Override
    public void startActivityForCls(@NonNull Class<? extends Activity> cls, @NonNull Serializable serializable) {
        Bundle bundle = new Bundle();
        bundle.putSerializable(getTag(), serializable);
        startActivityForCls(cls, bundle);
    }

    @Override
    public void startActivityForCls(@NonNull Class<? extends Activity> cls, @NonNull Parcelable parcelable) {
        Bundle bundle = new Bundle();
        bundle.putParcelable(getTag(), parcelable);
        startActivityForCls(cls, bundle);
    }

    @Override
    public void startActivityForCls(@NonNull Class<? extends Activity> cls, @NonNull Parcelable[] parcelables) {
        Bundle bundle = new Bundle();
        bundle.putParcelableArray(getTag(), parcelables);
        startActivityForCls(cls, bundle);
    }

    @Override
    public void postEvent(MessageEvent msgEvent) {
    }

    @Override
    public void register(Object subscriber) {
    }

    @Override
    public void unregister(Object subscriber) {
    }

    public void release() {
        if (handlePool != null)
            handlePool.release(this);
    }

    public static CommonHandler obtain(Context context) {
        CommonHandler commonHandler = handlePool.acquire();
        if (commonHandler != null) {
            commonHandler.init(context);
        } else {
            commonHandler = new CommonHandler(context);
        }
        return commonHandler;
    }


}

File: app/src/main/java/com/hlibrary/base/CommonHandler.kt
package com.hlibrary.base

import android.app.Activity
import android.app.ProgressDialog
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.Parcelable
import android.support.annotation.IntRange
import android.support.annotation.NonNull
import android.support.annotation.StringRes
import android.support.v4.content.LocalBroadcastManager
import android.support.v4.util.Pools
import android.text.TextUtils
import com.hlibrary.application.AppContext
import com.hlibrary.base.activity.BaseActivity
import com.hlibrary.util.Logger
import com.hlibrary.util.ToastUtil
import java.io.Serializable
import java.util.concurrent.Future

class CommonHandler {

    companion object {
        private val TAG = "CommonHandler"
        private val handlePool: Pools.SimplePool<CommonHandler> = Pools.SimplePool(100)
        private val defaultReqCode = -100;

        fun obtain(context: Context): CommonHandler {
            var commonHandler = handlePool.acquire()
            if (commonHandler != null) {
                commonHandler._init(context)
            } else {
                commonHandler = CommonHandler(context)
            }
            return commonHandler
        }
    }

    private var localBroadcastManager: LocalBroadcastManager? = null
    private var context: Context? = null
    var ExtraTag: String? = null
    private var progressDialog: ProgressDialog? = null

    constructor(context: Context) {
        _init(context)
    }

    private fun _init(context: Context): CommonHandler {
        this.context = context
        return this
    }

    fun pushActivity(activity: BaseActivity<*>) {
        AppContext.instance.activityManager.pushActivity(activity)
    }

    fun pullActivity(activity: BaseActivity<*>) {
        AppContext.instance.activityManager.pullActivity(activity)
    }

    fun getTag(): String {
        Logger.getInstance().i(TAG, " $TAG $ExtraTag")
        return if (!TextUtils.isEmpty(ExtraTag)) {
            ExtraTag!!
        } else TAG
    }

    fun showToast(text: String) {
        ToastUtil.showLongTime(context!!, text)
    }

    fun showToast(text: String, @IntRange(from = 1) duration: Int) {
        ToastUtil.showCustomTime(context!!, text, duration)
    }

    fun showToast(@StringRes resId: Int) {
        ToastUtil.showLongTime(context!!, resId)
    }

    fun showToast(@StringRes resId: Int, @IntRange(from = 1) duration: Int) {
        ToastUtil.showCustomTime(context!!, resId, duration)
    }


    fun showProgess(msg: String) {
        if (null == progressDialog) {
            progressDialog = ProgressDialog(context)
            progressDialog?.setMessage(msg)
            progressDialog?.setCancelable(false)
            progressDialog?.show()
        } else {
            progressDialog?.setMessage(msg)
            progressDialog?.show()
        }
    }


    fun dismissProgress() {
        if (progressDialog != null && progressDialog?.isShowing()!!)
            progressDialog?.dismiss()
    }


    fun runNewThread(r: Runnable): Future<*> {
        return AppContext.threadPool.submit(r)
    }


    fun runOnUiThread(r: Runnable) {
        if (context is BaseActivity<*>) {
            (context as BaseActivity<*>).runOnUiThread(r)
        } else {
            AppContext.mainHandler.post(r)
        }
    }


    fun runOnUiThreadForDelay(r: Runnable, @IntRange(from = 1) delayMillis: Long) {
        AppContext.mainHandler.postDelayed(r, delayMillis)
    }

    fun registerReceiver(receiver: BroadcastReceiver?, @NonNull vararg actions: String) {
        if (receiver == null || actions?.size == 0)
            return

        val filter = IntentFilter()
        for (action in actions) {
            filter.addAction(action)
        }
        localBroadcastManager?.registerReceiver(receiver, filter)
    }

    fun unregisterReceiver(r: BroadcastReceiver?) {
        if (r == null)
            return
        localBroadcastManager?.unregisterReceiver(r)
    }


    fun sendBroadcast(intent: Intent): Boolean {
        return localBroadcastManager?.sendBroadcast(intent)!!
    }

    @JvmOverloads
    fun startActivityForCls(cls: Class<out Activity>, req: Int = defaultReqCode): Boolean {
        if (req == defaultReqCode) {
            context?.startActivity(Intent(context, cls))
        } else {
            if (context is Activity) {
                (context as Activity).startActivityForResult(Intent(context, cls), req)
                return true
            }
        }
        return false
    }

    @JvmOverloads
    fun startActivityForCls(cls: Class<out Activity>, options: Bundle, req: Int = defaultReqCode): Boolean {
        if (req == defaultReqCode) {
            val intent = Intent(context, cls)
            intent.putExtras(options)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TASK
            context?.startActivity(intent)
            return true
        } else {
            if (context is Activity) {
                val intent = Intent(context, cls)
                intent.putExtras(options)
                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TASK
                (context as Activity).startActivityForResult(intent, req)
                return true
            }
        }
        return false
    }

    @JvmOverloads
    fun startActivityForCls(cls: Class<out Activity>, value: String, req: Int = defaultReqCode): Boolean {
        val bundle = Bundle()
        bundle.putString(getTag(), value)
        return startActivityForCls(cls, bundle, req)
    }

    @JvmOverloads
    fun startActivityForCls(cls: Class<out Activity>, values: Array<String>, req: Int = defaultReqCode): Boolean {
        val bundle = Bundle()
        bundle.putStringArray(getTag(), values)
        return startActivityForCls(cls, bundle, req)
    }

    @JvmOverloads
    fun startActivityForCls(cls: Class<out Activity>, serializable: Serializable, req: Int = defaultReqCode): Boolean {
        val bundle = Bundle()
        bundle.putSerializable(getTag(), serializable)
        return startActivityForCls(cls, bundle, req)
    }

    @JvmOverloads
    fun startActivityForCls(cls: Class<out Activity>, parcelable: Parcelable, req: Int = defaultReqCode): Boolean {
        val bundle = Bundle()
        bundle.putParcelable(getTag(), parcelable)
        return startActivityForCls(cls, bundle, req)
    }

    @JvmOverloads
    fun startActivityForCls(cls: Class<out Activity>, parcelables: Array<Parcelable>, req: Int = defaultReqCode): Boolean {
        val bundle = Bundle()
        bundle.putParcelableArray(getTag(), parcelables)
        return startActivityForCls(cls, bundle, req)
    }


    fun release() {
        handlePool.release(this)
    }

    fun getBundle(): Bundle? {
        if (context is Activity) {
            return (context as Activity).getIntent().getExtras()
        }
        return null
    }

    fun setResult(bundle: Bundle) {
        if (context is Activity) {
            val intent = Intent()
            intent.putExtras(bundle)
            (context as Activity).setResult(Activity.RESULT_OK, intent)
        }
    }

}